torch/csrc/jit/passes/memory_planning.cpp

```diff
@@ -0,0 +1,159 @@
+
+#include <c10/core/ScalarType.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/memory_planning.h>
+#include <torch/csrc/jit/runtime/static/impl.h>
+
+namespace torch {
+namespace jit {
+
+c10::TensorTypePtr checkIsSupported(const Value& value) {
+  TORCH_CHECK(
+      value.type()->cast<TensorType>(),
+      "out isn't a tensortype ",
+      *value.type());
+  auto ttp = value.type()->expect<TensorType>();
+  TORCH_CHECK(
+      ttp->scalarType().has_value(),
+      "This output was profiled but didn't have a scalar type.")
+  TORCH_CHECK(
+      ttp->sizes().concrete_sizes().has_value(),
+      "This output was profiled but doesn't have sizes.")
+  return ttp;
+}
+
+std::pair<std::vector<int64_t>, std::vector<int64_t>> getSizesStrides(
+    const c10::TensorTypePtr& ttp) {
+  auto sizes = ttp->sizes().concrete_sizes().value();
+  std::vector<int64_t> strides;
+  auto _strides = ttp->strides().concrete_sizes();
+  if (_strides.has_value() && _strides.value()[0] != 0) {
+    strides = _strides.value();
+  } else {
+    strides = at::detail::defaultStrides(sizes);
+  }
+  return std::make_pair(sizes, strides);
+}
+
+size_t computeStorageSize(const c10::TensorTypePtr& ttp) {
+  std::vector<int64_t> sizes, strides;
+  std::tie(sizes, strides) = getSizesStrides(ttp);
+  return at::detail::computeStorageNbytes(
+      sizes, strides, elementSize(ttp->scalarType().value()));
+}
+
+std::pair<size_t, std::map<Value*, c10::TensorTypePtr>>
+computeAlignedStorageSize(std::vector<Node*>& nodes) {
+  // TODO: liveness analysis
+  std::map<Value*, c10::TensorTypePtr> managed_tensors;
+  size_t managed_bytes = 0;
+  for (const auto& node : nodes) {
+    for (const auto& output : node->outputs()) {
+      auto ttp = checkIsSupported(*output);
+      managed_bytes += computeStorageSize(ttp);
+      managed_tensors.insert({output, ttp});
+    }
+  }
+
+  auto aligned_size = MemoryPlanner::compute_aligned_tensor_size(managed_bytes);
+  return std::make_pair(aligned_size, managed_tensors);
+}
+
+DeviceType findDeviceType(std::shared_ptr<Graph>& graph) {
+  for (const auto& inp : graph->inputs()) {
+    auto typ = inp->type()->cast<TensorType>();
+    if (typ && typ->device().has_value()) {
+      return typ->device().value().type();
+    }
+  }
+
+  for (const auto& node : graph->nodes()) {
+    for (const auto& inp : node->inputs()) {
+      auto typ = inp->type()->cast<TensorType>();
+      if (typ && typ->device().has_value()) {
+        return typ->device().value().type();
+      }
+    }
+  }
+  TORCH_CHECK(false, "couldn't find device type")
+}
+
+Node* insertAllocStorageNode(std::shared_ptr<Graph>& graph, size_t total_size) {
+  auto deviceType = findDeviceType(graph);
+  auto* storage = graph->create(prim::AllocateStorage, 1);
+  // TODO: pass device type here
+  storage->i_(attr::total_size, total_size);
+  storage->i_(attr::device, static_cast<int8_t>(deviceType));
+  storage->insertBefore(graph->nodes().front());
```

Comment:
I was wondering if we should pass these as arguments to `AllocateStorage` (as constants for now) so when we will get to supporting dynamic shapes, we would just compute the sizes (instead of passing constants)?

Reply:
naturally inputs vs attrs has its pros and cons: pro is what you said, con is it clutters the IR making downstream manipulation more difficult (i think?). i can the "input constant" route and compare differences.

Reply:
I wonder if we should lean into it in some later PR. I don't know how else we would support dynamic shapes and it would be nice to flush out any potential issues as early as possible as they can affect an overall design, IMO. @desertfire  @hlu1  ?